Introduce `SafelyIndex` to reach into data without panicking

felipesere opened the pull request.

Ethiraric commented:
Sorry I haven't had much time to dedicate to `saphyr`.

I'd like a couple changes on this:
  - Moving the traits definition to a separate file (`index.rs`)
  - Move `impl`s to `define_yaml_object_index_traits_impl`.
  
This guarantees that all YAML objects have the same interface.

felipesere commented:
I've updated the approach - using an enum rather than a trait implementation - which also made implementing the trait much simpler.
I've also added it to the macros, but I did realise that MarkedYaml* does not use the same macro expansion.
Curious if that is something you'd change down the line? For now I've implemented the `SafelyIndex` trait for them as that is what I mostly use over in my app.

felipesere commented:
Curious for your thoughts on this @Ethiraric 😄
